syntax = "proto3";

package raft;

// 定义 RequestVote RPC 请求消息
message RequestVoteRequest {
    int32 term = 1;
    int32 candidateId = 2;
    int32 lastLogIndex = 3;
    int32 lastLogTerm = 4;
}

// 定义 RequestVote RPC 响应消息
message RequestVoteResponse {
    int32 term = 1;
    bool voteGranted = 2;
}

// 定义 AppendEntries RPC 请求消息
message AppendEntriesRequest {
    int32 term = 1;
    int32 leaderId = 2;
    int32 prevLogIndex = 3;
    int32 prevLogTerm = 4;
    repeated LogEntry entries = 5;
    int32 leaderCommit = 6;
}

// 定义 AppendEntries RPC 响应消息
message AppendEntriesResponse {
    int32 term = 1;
    bool success = 2;
}

// 定义日志条目
message LogEntry {
    int32 term = 1;
    string command = 2;
}
// Raft 服务
service Raft {
    // AppendEntries 处理日志复制
    rpc AppendEntries(AppendEntriesRequest) returns (AppendEntriesResponse);
    
    // RequestVote 处理领导者选举
    rpc RequestVote(RequestVoteRequest) returns (RequestVoteResponse);
  }